### DISCUSSION TITLE
Rendering 2D images from GPU memory

### DISCUSSION THREAD

**Discussion by JDr111** [AvaloniaUI/Avalonia#14185] [Q&A] [answered]
Hello,
I'm developing a cross-platform app which needs to render 2D images. These images are to be constructed using hardware-accelerated means (e.g. CUDA for Nvidia devices and OpenCL for everyone else), therefore the resulting image data to be displayed on screen already resides in GPU memory. Both CUDA and OpenCL allow for sharing buffer data with DirectX and OpenGL. There's also the possibility of using Vulkan, both for CUDA interop and for doing the compute stuff itself instead of OpenCL. Performance is an important factor here, hence why GPUs and unmanaged C routines are used to accelerate constructing of the images. No complex 3D rendering is expected to be done here (at most rendering a single textured quad with alternating texture (the image data) if there's no better way).

I'm exploring the viability of using Avalonia for this app (haven't really used it before), and I do have a few questions.

I am aware of the GPU Interop API Avalonia offers. Assuming Windows as the host OS, one way of getting the CUDA/OpenCL buffer data into the Avalonia control should be by using the CUDA-DirectX (or CUDA-OpenGL) interop to obtain a texture object, which can then be imported into Avalonia through the GPU Interop API (so a "CUDA/OpenCL -> DirectX/OpenGL -> Avalonia" chain; the "CUDA/OpenCL" part is all unmanaged code for performance reasons).

- Although a bit messy, it hopefully should work? Is there any less complicated (but equally performant) means of importing the CUDA/OpenCL-generated 2D images into Avalonia?

One of the priorities for the project is to be cross-platform. That's where things get a bit hairy. Windows is primarily all about DirectX and DXGI interfaces (although OpenGL and Vulkan are usable just fine), Linux works primarily with OpenGL, though GPU drivers are getting there with Vulkan support as well, and Apple only supports its own Metal API in its newest hardware and OS releases. This means, in order to minimize code duplicity, the custom control for the project should be implemented with either OpenGL or Vulkan as the backend for Windows and Linux versions of the app, and with the Metal backend for the Mac OS X version (this would include using Metal for the image processing as well). Interop between CUDA and Vulkan isn't as straight-forward, but ultimately should work as well.

- What's your experience with using Avalonia's GPU Interop API on Linux? Does it work with OpenGL only, or Vulkan as well? Is it stable?
- What's the status of Avalonia's #12737 and its GPU interop capabilities?
- Are there plans to support the Metal API in Avalonia, along with GPU interop?

Thanks for any insights!

**Comment by kekekeks** [comment, marked as answer]
https://github.com/AvaloniaUI/Avalonia/issues/9925

Currently works for Windows and Linux (driver support may vary).

For metal there are plans for consuming IOSurface-backed textures.

**Reply by JDr111** [reply]
Awesome, thanks for the quick reply! I may opt for leaving Mac OS X support for a later time in that case.
